### PR TITLE
DL: Enable transfer learning

### PR DESCRIPTION
**Pull request by kaknikhil**
DL: Add model_weights, name and desc to the model arch table.

JIRA: MADLIB-1348

This commit adds three optional params to the model arch interface.
1. model weights in bytea format
2. name
3. description

The model_weights param will allow the user to load pre trained weights
to enable transfer learning

1. Remove image count from the serialized model weights that get stored
in the model output table. This image count was unnecessary and also
caused inconsistency in the model weights format in model_arch and
keras fit model out table.
2. Use bytea to store/read model weights. This includes the input param
and output col type. Here is why
  a. we found that in python, using double precision[] or real[] was
  almost double in size as compared to bytea.
  b. It also helps keep the internal state of the model and the output to
  the user consistent.

Additionally
1. Use plpy prepare to format the model weights as bytea.
2. Modify the deserialize code to accept a bytea string
instead of a double precision[]
3. Modify madpack code so that install check user can create python
UDFs in the madlib keras dev check sql.
4. Add dev check test to test for transfer learning by creating a UDF
that calls the load_arch_table with some pre defined weights so that the
first iteration of fit always returns the same loss and metric.
5. Rename serializer functions to be more reflective of their purpose.
6. Move model_shapes to wrapper and remove model_shapes from SD
7. Rename model_state to either state or serialized_weights depending on
its content.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/936/

File: src/ports/postgres/modules/deep_learning/keras_model_arch_table.py_in [hunk 1]
```diff
@@ -53,88 +53,83 @@ class Format:
...
+    select_query = "SELECT {model_id_col}, {model_arch_col} FROM {model_arch_table} " \
+                   "WHERE {internal_id_col} = '{unique_str}'".format(
+                    model_id_col=ModelArchSchema.MODEL_ID,
+                    model_arch_col=ModelArchSchema.MODEL_ARCH,
+                    model_arch_table=model_arch_table,
+                    internal_id_col=ModelArchSchema.__INTERNAL_MADLIB_ID__,
+                    unique_str=unique_str)
+    select_res = plpy.execute(select_query,1)
+
+    if len(select_res) != 1 or select_res[0][ModelArchSchema.MODEL_ARCH] != model_arch:
```
[the remark this answers is not in the document]
Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/keras_model_arch_table.py_in]: This was existing code that we chose not to change but we also thought of deleting this check. I will go ahead and delete it now.

File: src/ports/postgres/modules/deep_learning/keras_model_arch_table.py_in [hunk 2]
```diff
@@ -53,88 +53,83 @@ class Format:
...
         plpy.error("Keras Model Arch: Invalid keras model arch table {0},"
                    " missing columns: {1}".format(model_arch_table, missing_cols))
 
     sql = """
            DELETE FROM {model_arch_table} WHERE {model_id_col}={model_id}
-          """.format(model_arch_table=model_arch_table, model_id_col=Format.MODEL_ID,
+          """.format(model_arch_table=model_arch_table, model_id_col=ModelArchSchema.MODEL_ID,
                      model_id=model_id)
-    res = _execute(sql)
+    res = plpy.execute(sql, 0)
```
[the remark this answers is not in the document]
Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/keras_model_arch_table.py_in]: I believe the reason here is that we pass 0 for all the queries that don't need to return any rows. I don't think it matters either way but I will keep the code for now.

File: src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in [hunk 3]
```diff
@@ -779,3 +779,93 @@ SELECT madlib_keras_predict(
...
+    'test_keras_model_arch_table',
+    1,
+    $$ optimizer=SGD(lr=0.001, decay=1e-6, nesterov=True), loss='categorical_crossentropy', metrics=['mae']$$::text,
+    $$ batch_size=2, epochs=1, verbose=0 $$::text,
+    3);
+SELECT training_loss_final, training_metrics_final FROM keras_saved_out_summary;
+
+--assert training loss and metric deterministic
+SELECT assert(abs(training_loss_final - 0.64) < 0.01 AND
+              abs(training_metrics_final - 0.47) < 0.01,
```
[the remark this answers is not in the document]
Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in]: +1 for the freezing layers assert.
I do think that there is value in asserting both weights and the loss/metric values. For now I will add a TODO comment to the dev check assert to try out the freezing layers idea so that we can do this in a future JIRA.

File: src/madpack/madpack.py [hunk 4]
```diff
@@ -1027,8 +1027,9 @@ def run_install_check(args, testcase, madpack_cmd):
         _internal_run_query("DROP OWNED BY %s CASCADE;" % (test_user), True)
         _internal_run_query("DROP USER IF EXISTS %s;" % (test_user), True)
 
-    _internal_run_query("CREATE USER %s;" % (test_user), True)
+    _internal_run_query("CREATE USER %s WITH SUPERUSER NOINHERIT;" % (test_user), True)
     _internal_run_query("GRANT USAGE ON SCHEMA %s TO %s;" % (schema, test_user), True)
+    _internal_run_query("GRANT ALL PRIVILEGES ON DATABASE %s TO %s;" % (db_name, test_user), True)
```
[the remark this answers is not in the document]
Comment (kaknikhil) [reply src/madpack/madpack.py]: These permissions only apply to the dev-check/install-check user who is created only for running the dev-check/install-check tests. This user is dropped at the end of each madpack run of dev-check/install-check.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/943/

File: src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in [hunk 5]
```diff
@@ -16,68 +16,44 @@
...
-    # get_image_count_from_state() we defined below, we should check if
-    # we still need to return it here when refactoring
-    return float(state[0]), model_weights
-
-def get_image_count_from_state(model_state):
-    if not model_state:
-        return None
-    state = np.fromstring(model_state, dtype=np.float32)
-    return float(state[0])
+def get_serialized_weights_from_state(state):
```
[the remark this answers is not in the document]
Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in]: added a section named `workflow` at the top of the serializer and renamed the functions as well

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/944/
